### ISSUE TITLE
Weird [mis]interaction between generics and union

### ISSUE THREAD

**Issue by ANogin** [python/mypy#17756]
**Bug Report**

In the following file
``` python
from typing import Generic, Sequence, Tuple, TypeVar

def f(
    x: Sequence[str], y: Sequence[int] | Sequence[bool]
) -> zip[Tuple[str, int]] | zip[Tuple[str, bool]]:
    return zip(x, y) # OK

T = TypeVar("T")

class C(Generic[T]):
    def g(
        self, x: Sequence[str], y: Sequence[T] | Sequence[bool]
    ) -> zip[Tuple[str, T]] | zip[Tuple[str, bool]]:
        return zip(x, y) # ERROR
```

mypy accepts `f`, but complains 
```
error: Incompatible return value type (got "zip[tuple[str, object]]", expected "zip[tuple[str, T]] | zip[tuple[str, bool]]")  [return-value]
```
on the return from `C.g`

Confirmed with mypy 1.11.2 in Python 3.11.9, but does not seem version-specific.

P.S. My specific use case where I found this issue was where I had `for (a,b) in zip(x,y)`, then was using `b` where `Union[T|bool]` was expected, and getting an error message from mypy that `object` (`b`) is used where `T|bool` was expected; I tried to simplify that code to digest it to the minimal reproducible example. I do not fully understand all the constraints here, but the fact that `f` is accepted and `C.g` is not does suggest to me that something is clearly wrong.

**Comment by A5rocks** [comment]
```py
def f(
    x: Sequence[str], y: Sequence[bytes] | Sequence[bool]
) -> zip[Tuple[str, bytes]] | zip[Tuple[str, bool]]:
    return zip(x, y) # does not work
```

and that's just replacing `int` with `bytes`. I haven't really thought through the types but I think your `C.g` is incorrect.

**Comment by ANogin** [comment]
Interesting. At the same time

```python
def f(x: Sequence[str], y: Sequence[bytes] | Sequence[bool]) -> zip[Tuple[str, bytes | bool]]:
    return zip(x, (e for e in y))
```

does work and is probably is an OK workaround...
